### ISSUE TITLE
Error during ingestion process when sharing GitHub app with multiple organizations

### ISSUE THREAD

**Issue by ProticM** [backstage/backstage#8463]
Multiple GitHub org reading fails when sharing GitHub app integration. The same GitHub app is installed on two orgs, integration is set to use only this single app, and reading fails for both organizations.

Is sharing a single GitHub app integration for multiple orgs supported or do we need to set up GitHub app per org and add each of them to the integration apps list?

## Expected Behavior
GitHub application installed on two orgs should provide access and enable processing of each GitHub org teams and users.

## Current Behavior
Ingestion process fails with error: `A JSON web token could not be decoded`.

## Steps to Reproduce

1. Create GitHub app
2. Install the app on two GitHub orgs
3. Set up GitHub integration in the config
4. Add `github-org` or `github-multi-org` processor
5. Run the app

## Context
We have multiple GitHub organization that we want to populate our catalog from, and we are using GitHub app integration. Our idea is to have a single GitHub app installed on each GitHub organization that we want to read from.

Is this supported or we need to have 1:1 relation (each org must have GitHub app and each app must be listed in the apps list in the integrations config)?

## Your Environment
- NodeJS Version (v14): v14.17.5
- Operating System and Version (e.g. Ubuntu 14.04): Windows 10
- Browser Information: Chrome

**Comment by benjdlambert** [comment]
I'm wondering if it's something to do with the `installationId` not being passed correctly and being set. It should work as expected, you should be able to share the GithubApp. Could you confirm that you have an `installationId` set in config?

**Comment by ProticM** [comment]
I can also mention that with our other app that we have, but installed only on one org, we do not have any problems with the same configuration. As soon as we switch to the one being installed on multiple orgs, the issue occurs.

In the config we have the following set:

`appId`
`clientId`
`clientSecret`
`webhookSecret`
`privateKey`

Are you referring to the `appId`? If so, we have this property set.

Correct me if I'm wrong, but installation ids are different after installing the same app across orgs and also, I did not find anywhere the information that we can set the `installationId` in the config 🤔

**Comment by benjdlambert** [comment]
Hmm no, I'm wondering if this bit is failing somehow. https://github.com/backstage/backstage/blob/a15d02851747791d6e2fde8f82d18b12cb990b85/packages/integration/src/github/GithubCredentialsProvider.ts#L91

**Comment by benjdlambert** [comment]
Could you maybe add some logging and give it a try and see what you find. Otherwise I can try and replicate this sometime this week.

**Comment by ProticM** [comment]
Yes, this was my next step, and that exact place I was targeting as a possible problem. Will see what I can do regarding logs, but feel free, once you get the time, to try to replicate it

**Comment by ProticM** [comment]
@benjdlambert It seems that it fails while listing the installations here: https://github.com/backstage/backstage/blob/a15d02851747791d6e2fde8f82d18b12cb990b85/packages/integration/src/github/GithubCredentialsProvider.ts#L138 and the error log is:

```js
 RequestError [HttpError]: A JSON web token could not be decoded
     at ...\backstage\node_modules\@octokit\request\dist-node\index.js:86:21
     at processTicksAndRejections (internal/process/task_queues.js:95:5)
     at async hook (...\backstage\node_modules\@octokit\auth-app\dist-node\index.js:380:18)
     at async Object.next (...\backstage\node_modules\@octokit\plugin-paginate-rest\dist-node\index.js:62:26) {
   status: 401,
   response: {
     url: 'https://api.github.com/app/installations',
     status: 401,
     headers: {
       // removed for brevity
     },
     data: {
       message: 'A JSON web token could not be decoded',
       documentation_url: 'https://docs.github.com/rest'
     }
   },
   request: {
     method: 'GET',
     url: 'https://api.github.com/app/installations',
     headers: {
       accept: 'application/vnd.github.v3+json',
       'user-agent': 'octokit-rest.js/18.5.6 octokit-core.js/3.2.4 Node.js/14.17.5 (win32; x64)',
       authorization: 'bearer [REDACTED]'
     },
     request: { hook: [Function: bound bound register] }
   }
 }
```

I see that the status is 401, I will double-check as well on our end again because as mentioned, with the other app installed only on one org it works fine.

But maybe the JWT token was not generated properly?

**Comment by jhaals** [comment]
Is this running on github.com or github enterprise? I'm not sure that apps can be shared between multiple organizations on github.com unless they're made public

**Comment by benjdlambert** [comment]
@ProticM so I've just tried to reproduce this again today, and can install an app after it's made public into mutliple places and be able to read locations from them from the catalog processors. What is it you're trying to do when you get this error.

Steps:
- use `yarn backstage-cli create-github-app org1`
- setup in app config and remove the `GITHUB_TOKEN`.
- create a test private repo in org1 and add that to `locations` in config
- go the app settings page and make it public and install that app into org2
- add a location for a private repo in org2

Does this bit work for you? Maybe it's specifically related to the org processors and maybe the github apps don't have the right permissions?

Just want to see if we're on the same page before I start digging into the org readers.

**Comment by benjdlambert** [comment]
Reason im asking is because that method you linked to will get called everywhere so it would be weird if it was throwing errors just in the org readers.

**Comment by benjdlambert** [comment]
Just confirmed that if I use a location with `type: github-org` and location as one of the organisations that the app is installed into, it first gave me the error about the resource not being accessible by the integration, because of the permissions. When updating the integration with the permissions theres an email sent to owners of each organisation to update. Once I updated that it worked fine for me. Is there anything you can think of that you're doing different that could help debug this?

**Comment by ProticM** [comment]
@jhaals Yes, it is running on enterprise account.

@benjdlambert Yes, you are correct, the error is everywhere which I can confirm now.

I've found out what the problem was. The problem was in the corrupted private key that is used to sign the token. After replacing the private key, the error `A JSON web token could not be decoded` is gone. I'm able to fetch the installations for all orgs using the new app that we created 👍 

However, on the new app, we are missing (I assume) the read-only access for members because now the error is `Resource not accessible by integration`. All other access rights are there.

We had the same problem (read-only access for members) with the old app that has no issues with the private key.

So, long story short, I can say that the error was on our end and that we can close this issue.

But, I still did not confirm that it will fetch the data for all orgs after applying the missing access rights. I assume it will fetch it without issues.

**Comment by benjdlambert** [comment]
Ah perfect - that's good to know. Glad it's all working out. Going to close this issue, and feel free to re-open if it doesn't work after you've added the org:read and member:read permissions I think you'll need, so that you can ingest emails and stuff like that too. :)

**Comment by ProticM** [comment]
@benjdlambert Hi, I have a quick question just to have the info about this. Can we integrate multiple GitHub apps to be used by different GitHub organizations? Is backstage capable of understanding which GitHub app is responsible for which GitHub org from the integrations config?

**Comment by benjdlambert** [comment]
I believe there is some logic to do this yes, you can provide an array of github apps for a host, and I think it will look up which app has the right perms for that path? I can't remember fully, but would be a good feature if it's not already supported.

**Comment by ProticM** [comment]
Yes, it would be beneficial to have that, I agree. I'm not sure, I did not analyze fully that part.

For example, we have one app that we share across our orgs and the problem we are facing is having a different set of permissions per org.
